Add liens to Terraform.

ndmckinley opened the pull request:
This resource is nonstandard in a number of ways.  This is a longish
read, settle in.  :)

First, it doesn't support `GET`s.  Instead, it supports a List operation.  This is
not unprecedented in MagicModules - we already have support for it in
Ruby, Puppet, and Ansible, and it's used for SQL Users, Flags and
Tiers, and DNS RecordSets.  This is what the `self_link_query` is for.

Second, it uses a different URL for `GET` and `DELETE`.  This *is*
unprecedented in MagicModules, and so I have done a hack - I used custom
code to rewrite the URL in `DELETE` before we send the request.  This
is not a good solution and I'm not happy with it.  Please suggest a better
one if you think it's worth it.

Third, it doesn't use `Operation`s at all.  This isn't unprecedented
in MagicModules - it just means we don't support timeouts and we
don't wait on operations.  But unfortunately we haven't written that
into Terraform, so I did that.

Fourth, it doesn't have a project.  We had some "no project" code in the
template preliminarily - we already didn't generate the "project" schema
entry if there was no project - but we still call `getProject` a lot and
that needed to be fixed.

Fifth, it doesn't have a self_link (because it doesn't support
`GET`).  We had some preliminary support for that, but it was broken
in the same way as the project support, and that needed to be fixed
as well.

Sixth, it has query parameters!  Those aren't unprecedented in magic
modules at all, you just handle them by making the self_link have the
necessary parameters, but currently terraform unconditionally concats
`?alt=json` to all our queries, and that's a bummer if there's already a
query parameter on there, so I fixed that.

Seventh, it is *very* finicky about how it accepts parameters - it
needs its project prefixed with `project/`, for instance, and it
only accepts project numbers instead of project names, and it stores
its name with the `liens/` prefix, which we generally prefer to
avoid.  I haven't fixed any of this at all - I think custom
expanders and flatteners are the right way, but I don't have a good
notion of how to build those reusably yet.  I'm still thinking on
that.  Resource works okay with that finickiness.

Eighth, resourceManager APIs are called all kinds of names.  It's called
`cloudresourcemanager` in the go generated code.  It's called
`gresourcemanager` in chef and puppet and ansible, which leads to
a generated name of `Resourcemanager` in Terraform, and that's wrong.
In general, it's messy.  So I added a shoddy hack to the top of
`resource.erb` to handle that.  Please suggest a better hack.

And finally, of course, I had to add in resourcemanager to be generated
by The Magician.

This is meant to address https://github.com/terraform-providers/terraform-provider-google/issues/1435.

-----------------------------------------------------------------
# [all]
Changes tangential to adding Liens resource (not yet supported here).
## [terraform]
Add liens resource.

Fixes #1435.
## [puppet]
### [puppet-dns]
### [puppet-sql]
### [puppet-compute]
## [chef]

modular-magician commented:
I am a robot that works on MagicModules PRs!

I built this PR into one or more PRs on other repositories, and when those are closed, this PR will also be merged and closed.
depends: https://github.com/terraform-providers/terraform-provider-google/pull/1484

ndmckinley commented:
@nelsonjr: I'd like your opinion on points number two and eight and `api.yaml`.
@danawillow and @rosbo: I'd like your opinion on everything else.  :)

rosbo commented on products/resourcemanager/terraform.yaml:
````diff
@@ -0,0 +1,53 @@
+      ```
+
+    custom_code: !ruby/object:Provider::Terraform::CustomCode
+      post_create: templates/terraform/lien_post_create.erb
````
For my other PRs using custom code, I have been following a pattern where I put the files in a directory inside templates based on the custom_code injection point. 
https://github.com/GoogleCloudPlatform/magic-modules/tree/master/templates/terraform/custom_flatten
https://github.com/GoogleCloudPlatform/magic-modules/tree/master/templates/terraform/constants
https://github.com/GoogleCloudPlatform/magic-modules/tree/master/templates/terraform/update_encoder

rosbo commented on templates/terraform/modify_delete_url.erb:
```diff
@@ -0,0 +1,4 @@
+url, err = replaceVars(d, config, "<%= object.__product.default_version.base_url -%>{{name}}")
```
Consider adding a comment here with an explanation about why you added this block. I would take the explanation you added in this PR description. The generated code is a bit confusing to read the way it is.

rosbo replied:
Another option is adding a `delete_url` and `create_url` to the Resource class.
If set `delete_url` and `create_url` are nil, default to the `base_url`.
Update the template to use these new urls.

rosbo commented on templates/terraform/provider_gen.erb:
```diff
@@ -18,6 +18,11 @@ package google
 
+
+<% if product_ns == 'Resourcemanager'
+		product_ns = 'ResourceManager'
```
Beautiful 🤦‍♂️

rosbo replied:
In a different PR, we might evaluate adding a way to override a product name the same way we do it for properties or resources.

nelsonjr replied:
You may want to extract the _meaning_ of the if into a concept and add it to `terraform.yaml`. Adding a product specific like this in the template is a major no-no.

rosbo commented on templates/terraform/resource.erb:
```diff
@@ -160,15 +181,60 @@ func resource<%= resource_name -%>Read(d *schema.ResourceData, meta interface{})
   }
 <% end -%>
+
+<% unless object.self_link_query.nil? -%>
```
I would move this to a separate template file and call `compile_template`

modular-magician commented:
I am (still) a robot that works on MagicModules PRs!

I just wanted to let you know that your changes (as of commit 329fb20d) have been included in your existing downstream PRs.

modular-magician commented:
I am (still) a robot that works on MagicModules PRs!

I just wanted to let you know that your changes (as of commit bbd4dac1) have been included in your existing downstream PRs.

modular-magician commented:
I am (still) a robot that works on MagicModules PRs!

I just wanted to let you know that your changes (as of commit b5aab98f) have been included in your existing downstream PRs.

modular-magician commented:
I am (still) a robot that works on MagicModules PRs!

I just wanted to let you know that your changes (as of commit 9b72b0ad) have been included in your existing downstream PRs.

ndmckinley commented:
FWIW: I think this is the best I can do for this resource right now.  Import, read, create, delete all work.  Some of it's a bit ugly and hacky but I think it's all correct.  There are two hack-fixes I can do:
* Make a 'name override' for products.
* Make `create_url` and `delete_url` and set them to `base_url` if nil.

Which, if either, do you think is worth doing in this PR?

rosbo commented:
I would submit it as-is and make the change for product override and create/delete urls in separate PRs

nelsonjr commented on products/resourcemanager/api.yaml:
```diff
@@ -93,3 +93,62 @@ objects:
             name: 'id'
             description: Id of the organization
 
+  - !ruby/object:Api::Resource
```
Suggestion: To make sure a new `api.yaml` works as expected we use to add to 2 providers at least. Even if they are separate PRs it is worth making sure nothing is inherently provider specific here.

nelsonjr commented on templates/terraform/resource.erb:
```diff
@@ -19,11 +19,22 @@ package google
+	# issues, so we just handle them here in the template.  If we start seeing this
+	# issue more than just in the ResourceManager APIs, we should invent a more general
+	# solution.
+	if product_ns == 'Resourcemanager'
```
ditto

nelsonjr commented:
@ndmckinley regarding top PR list:

Two: That's okay when you have create and delete that are custom. For DNS we have a custom code for create that's different from delete.

Eight: One thing we could do... retire prefix in favor of a `short_name`. In the short name you have "Compute", "Resource Manager", etc. In P/C change the prefix method to build "g#{short_name.lower.no_spaces}". Which can be overridden later if needed. (although override does not affect non-resources [like product] today it should be easy to extend)

ndmckinley commented:
@nelsonjr 
2)  Glad to hear it.  Going to stick with that, then, since it's the thing that's done elsewhere in MM, unless we end up seeing more of them than you'd expect.

8)  Good idea ... but I think this PR already fixes way too many things at once.  I'm getting uncomfortable continuing to add on.  I agree to make my next 2 tasks "remove the gross Resourcemanager -> ResourceManager hack" and "Add liens to puppet".

@rosbo @danawillow 
I fixed up number seven in the PR description (less reusably than I wanted, but it's the best I can do), and added on a number ten: You have to import this resource with its parent, and "parent" != "project", so I had to add the ability to use custom import IDs.

ndmckinley commented:
Sure!

modular-magician commented:
I am (still) a robot that works on MagicModules PRs!

I just wanted to let you know that your changes (as of commit 0dbcf106) have been included in your existing downstream PRs.

nelsonjr commented:
That's reasonable. I suggest you file an issue so it does not get forgotten and TODO items in the code in the right places with the link to the issue.

ndmckinley commented on products/resourcemanager/terraform.yaml:
```diff
@@ -0,0 +1,53 @@
+      post_create: templates/terraform/lien_post_create.erb
+      pre_delete: templates/terraform/modify_delete_url.erb
+# This is for copying files over
+files: !ruby/object:Provider::Config::Files
```
Sure thing, done.

ndmckinley commented on templates/terraform/decoders/avoid_meaningless_project_update.erb:
```diff
@@ -0,0 +1,54 @@
+  // conditions are:
+  // 1) if the new or old values contain '/', the prefix of that is 'projects'.
+  // 2) if either is non-numeric, a project with that ID exists.
+  // 3) the project IDs represented by 
```
Oh, hm, are we supposed to finish our sentences?  I'll keep that in mind for when I

danawillow commented:
@nelsonjr @rosbo I think we should all agree where these custom code templates go- right now this PR has one in products/resourcemanager/helpers and others in templates/terraform/[post_import, pre_delete, etc.]

modular-magician commented:
I am (still) a robot that works on MagicModules PRs!

I just wanted to let you know that your changes (as of commit 76c19434) have been included in your existing downstream PRs.

modular-magician commented:
I am (still) a robot that works on MagicModules PRs!

I just wanted to let you know that your changes (as of commit c1fb7dbe) have been included in your existing downstream PRs.

modular-magician commented:
I am (still) a robot that works on MagicModules PRs!

I just wanted to let you know that your changes (as of commit 96a8b7a1) have been included in your existing downstream PRs.

modular-magician commented:
I am (still) a robot that works on MagicModules PRs!

I just wanted to let you know that your changes (as of commit e6de95c6) have been included in your existing downstream PRs.

rileykarson commented:
Hey @ndmckinley did we lose the tests here? They are in Terraform but not MM.

ndmckinley commented:
Yep, that's right - didn't include the tests in autogeneration.  At the time we didn't know we were going to need to autogenerate all the tests eventually ... or at least I didn't.  :)
